Add IProvideEnabledStateBackend interface

come-nc opened the pull request:
This allows user backends to present some users as disabled, and to manage themselves how disabling users works.
It is used in the LDAP backend to support a new option to make remnants users appear as disabled.

- [x] The configuration option is not exposed in the UI

come-nc committed:
Add user enabled state backend feature

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Use the new IProvideEnabledStateBackend interface in user_ldap

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Rebased and fixed conflicts.

@blizzz Should the configuration be global like implemented or for each ldap configuration? If we want to put in the UI we have to set it as a per-configuration setting anyway, no?

blizzz commented:
> @blizzz Should the configuration be global like implemented or for each ldap configuration? If we want to put in the UI we have to set it as a per-configuration setting anyway, no?

We can also add another section with global LDAP settings. We have a few of them, but none are exposed. Is there a scenario where different settings per config make sense?  For me it makes more sense as a global flag, otherwise we are inconsistent across configs, potentially resulting in confused admins.

come-nc commented:
> We can also add another section with global LDAP settings. We have a few of them, but none are exposed. Is there a scenario where different settings per config make sense? For me it makes more sense as a global flag, otherwise we are inconsistent across configs, potentially resulting in confused admins.

Well I don’t know, a per-configuration setting can always be configured the same way for all configurations while if we make it a global setting people are stuck with it. But I have to say I do not have a particular usecase in mind.
If we make it global, can we leave it as-is and keep adding a config UI for a following PR?

Do we have a list of global settings that would make sense to appear in that UI?

blizzz commented:
> Do we have a list of global settings that would make sense to appear in that UI?

By the power of `ack` we have in app config:

- enforce_home_folder_naming_rule (i don't recall this one :sweat_smile: )
- cleanUpJobChunkSize (doc should be sufficient)
- bgjRefreshInterval (doc should be sufficient)

For completeness:
- cleanUpJobOffset (controlled by the app, background job state)
- background_sync_interval (controlled by the app, background job state)
- background_sync_prefix (controlled by the app, background job state)
- background_sync_offset (controlled by the app, background job state)


Then we also have system keys (config.php):

- ldapProviderFactory (do not expose)
- ldap_log_file (would be a candidate)
- ldapUserCleanupInterval (doc should be sufficient)

come-nc committed:
[user_ldap] Add per-connection setting for marking remnants as disabled

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Rebased and added the option to the UI.
I added it per-connection, because this is how the UI is organized, it fits nicely next to the other options in advanced.

come-nc committed:
Fix user tests, avoid setting enabled state to the same value

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix autoloaders

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Use DI for DeletedUsersIndex and fix tests

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix User_Proxy tests as well

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

artonge commented on lib/public/User/Backend/IProvideEnabledStateBackend.php:
```diff
@@ -0,0 +1,47 @@
+	 * @param callable():bool $queryDatabaseValue A callable to query the enabled state from database
+	 * @param callable(bool):void $setDatabaseValue A callable to set the enabled state in the database.
+	 */
+	public function setUserEnabled(string $uid, bool $enabled, callable $queryDatabaseValue, callable $setDatabaseValue): void;
```
```suggestion
	public function setUserEnabledState(string $uid, bool $enabled, callable $queryDatabaseValue, callable $setDatabaseValue): void;
```

come-nc replied:
```suggestion
	public function setUserEnabled(string $uid, bool $enabled, callable $queryDatabaseValue, callable $setDatabaseValue): bool;
```

artonge commented on lib/public/User/Backend/IProvideEnabledStateBackend.php:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @since 26.0.0
+ */
+interface IProvideEnabledStateBackend {
```
```suggestion
interface IProvideUserEnabledStateBackend {
```

come-nc replied:
It’s already in a User namespace and other backends do not include user in their names.

come-nc committed:
Fix since annotations and add boolean return type for setUserEnabled

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
So, since I started working on this I discovered we have no method to list disabled users.
As we need to avoid doing so by calling isEnabled on all users, I intend to add a method to this new interface to list disabled users.

come-nc committed:
Add method to list disabled users to IProvideEnabledStateBackend

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
> So, since I started working on this I discovered we have no method to list disabled users. As we need to avoid doing so by calling isEnabled on all users, I intend to add a method to this new interface to list disabled users.

Done.

Fenn-CS commented on apps/user_ldap/lib/User_LDAP.php:
```diff
@@ -669,4 +668,21 @@ public function createUser($username, $password) {
+		return $enabled;
+	}
+
+	public function getDisabledUserList(int $offset = 0, ?int $limit = null): array {
+		throw new \Exception('This is implemented directly in User_Proxy');
+	}
```
By the way

- Is the intention that we never want people to call this method directly from `User_LDAP` and instead call it from `User_Proxy`
- Or that this is still to be implemented?

This should probably have more commentary.

come-nc replied:
The first solution, this will always be called directly on User_Proxy.
For other methods User_Proxy proxies the call to User_LDAP, but not for this one.

Pytal commented:
/backport to stable26
